### PR TITLE
feat: add score column in spotlighttable

### PR DESCRIPTION
**Pull request by zhengow**
**Thank you for your contribution** to the [KodaDot Indexer](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

### What's new?

- [ ] PR closes #<issue_number>
- [x] This PR is required for kodadot/nft-gallery#3001

### Before submitting Pull Request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [x] I've posted a screenshot of demonstrated change in this PR

### Optional

- [x] I've run the indexer and it hasn't failed
- [ ] I've changed schema and performed a migration
- [ ] I found edge cases

### Screenshot

- [ ] My contribution is a **resolver** so I have pic from playground

image
image

### CODE REVIEW

File: src/server-extension/resolvers/spotlight.ts [hunk 1]
```diff
@@ -24,28 +26,47 @@ enum OrderDirection {
...
+    try {
+      tmpPassionList = JSON.parse(passionList)
+      if (!Array.isArray(tmpPassionList)) {
+        throw new Error("not array")
+      }
+      if (tmpPassionList.some(passion => typeof passion !== "string")) {
+        throw new Error("not string array")
+      }
+    } catch {
+      console.error("Invalid passionList")
```
Comment (petersopko) [inline src/server-extension/resolvers/spotlight.ts:47]: hey, I'm wondering how should this work, I'm passing in various addresses and still ending with "Invalid passionList" in console

**Comment by petersopko** [comment]
sorting seems to work, pls check the error :)

**Review by petersopko — CHANGES_REQUESTED** [review]

**Comment by vikiival** [comment]
Please be aware that spotlight resolver should not be used.

**Comment by zhengow** [comment]
> Please be aware that spotlight resolver should not be used.

why is it?

**Comment by vikiival** [comment]
> why is it?


we should be using the cached version of it.  

Please see:
https://github.com/kodadot/rubick/blob/4579f5007f92e8debf8dd579a83fe37691ded1dd/src/mappings/utils/cache.ts

**Comment by zhengow** [comment]
> > why is it?
> 
> we should be using the cached version of it.
> 
> Please see: https://github.com/kodadot/rubick/blob/4579f5007f92e8debf8dd579a83fe37691ded1dd/src/mappings/utils/cache.ts

done. Do I need to commit generated files: `src/model/generated/spotlight.model.ts` and `db/migrations/1656868613783-Data.js`
image

**Comment by petersopko** [comment]
> > we should be using the cached version of it.
> > Please see: https://github.com/kodadot/rubick/blob/4579f5007f92e8debf8dd579a83fe37691ded1dd/src/mappings/utils/cache.ts
> 
> Done. Do I need to commit generated files: `src/model/generated/spotlight.model.ts` and `db/migrations/1656868613783-Data.js` ?

I've done the testing, this wfm, can't give a definite answer about the generated files here, my guess is yes, so the migration doesn't have to be done every time in the future, @vikiival can you please help here?

**Review by petersopko — APPROVED** [review]

**Review by petersopko — CHANGES_REQUESTED** [review]
ok, after discussing this, you need to commit the generated files as well, in order to have the migration recorded in the repo

**Comment by petersopko** [comment]
@zhengow that one last thing and it's good to go :)

**Comment by zhengow** [comment]
> @zhengow that one last thing and it's good to go :)

great! Already commit files.😀

**Comment by zhengow** [comment]
> @zhengow that one last thing and it's good to go :)

When can we merge and move on🙌

File: src/mappings/utils/cache.ts [hunk 2]
```diff
@@ -44,7 +44,8 @@ enum Query {
         COUNT(*)                                                       as total,
         COUNT(distinct ne.current_owner)                               as unique_collectors,
         SUM(CASE WHEN ne.issuer <> ne.current_owner THEN 1 ELSE 0 END) as sold,
-        COALESCE(SUM(e.meta::bigint), 0)                               as volume
+        COALESCE(SUM(e.meta::bigint), 0)                               as volume,
+        SUM(CASE WHEN ne.issuer <> ne.current_owner THEN 1 ELSE 0 END) * COUNT(distinct meta_id) / COUNT(*) as score
```
Comment (vikiival) [inline src/mappings/utils/cache.ts:48]: Please do not use `*` 

It's not performant wise

**Review by vikiival — CHANGES_REQUESTED** [review]
smol stuff

**Comment by zhengow** [comment]
Hi， @vikiival 🤲
Cloud you pls check thid？

**Comment by vikiival** [comment]
> Cloud you pls check this？

sir there is a unresolved comment

**Comment by zhengow** [comment]
> > Cloud you pls check this？
> 
> sir there is a unresolved comment

Already change `count(*)` to `count(1)`, how about now?

**Comment by zhengow** [comment]
@vikiival Do u have time to check this👀

File: schema.graphql [hunk 3]
```diff
@@ -142,6 +142,7 @@ type Spotlight @entity {
   total: Int!
   average: Float
   volume: BigInt
+  score: Float!
```
Comment (vikiival) [inline schema.graphql:145]: Please remove `!`

**Review by vikiival — CHANGES_REQUESTED** [review]

**Review by vikiival — CHANGES_REQUESTED** [review]
https://github.com/kodadot/rubick#development-flow

**Comment by zhengow** [comment]
Sorry I forget this. First time to make PR for db. Already done `codegen` work.
How about add `codegen` command in husky that everyone can avoid this 👀

**Comment by vikiival** [comment]
> How about add codegen command in husky that everyone can avoid this

Feel free to make a PR love to have that!

**Comment by zhengow** [comment]
Let me check this later👌






------------------ Original ------------------
From: Viki Val ***@***.***&gt;
Date: Fri,Jul 29,2022 3:23 AM
To: kodadot/rubick ***@***.***&gt;
Cc: zhengow ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [kodadot/rubick] feat: add score column in spotlighttable (PR#107)





  
How about add codegen command in husky that everyone can avoid this
  
Feel free to make a PR love to have that!

**Comment by vikiival** [comment]
Last thing you are missing is to update the migration and I think we are good to go

**Comment by zhengow** [comment]
> Last thing you are missing is to update the migration and I think we are good to go

The step to doing is:
1. checkout to branch `main`
2. `just up` to start db
3. `just reset` to init db
4. checkout to my branch
5. `just update-db` to get migration file

Am I right?
However, it seem that the file that I get is just the same as [4578a](https://github.com/kodadot/rubick/pull/107/commits/4578a4219a969e63c095499ebe2350964512227e)

**Comment by petersopko** [comment]
@zhengow this one might be helpful? https://medium.com/kodadot/index-your-kusama-nfts-b4fab5ae1628

**Comment by zhengow** [comment]
> Last thing you are missing is to update the migration and I think we are good to go

update migration file, remove `not null` words manually. Now is it correct?
But why the automatic generated file has `not null` filed? 👀

**Comment by vikiival** [comment]
The PR looks legit can you please run it again to be sure that everything works?

**Comment by vikiival** [comment]
Closing as obsolete, thanks for the contribution

**Pull request closed without merging**